File: crates/nba_core/src/schema.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct League {
    pub league_id: i32,
    pub league_code: String,
    pub league_name: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Season {
    pub season_id: i32,
    pub league_id: i32,
    pub season_year: i32,
    pub start_date: Option<String>,
    pub end_date: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Arena {
    pub arena_id: i32,
    pub arena_name: String,
    pub city: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Team {
    pub team_id: i64,
    pub league_id: i32,
    pub franchise_code: Option<String>,
    pub nba_api_team_id: Option<i64>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TeamHistory {
    pub team_history_id: i32,
    pub team_id: i64,
    pub effective_start: String,
    pub effective_end: Option<String>,
    pub city: String,
    pub nickname: String,
    pub abbreviation: Option<String>,
    pub is_active: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Player {
    pub player_id: i64,
    pub nba_api_person_id: Option<i64>,
    pub first_name: Option<String>,
    pub last_name: Option<String>,
    pub display_name: Option<String>,
    pub birth_date: Option<String>,
    pub from_year: Option<i32>,
    pub to_year: Option<i32>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Coach {
    pub coach_id: String,
    pub nba_api_coach_id: Option<i64>,
    pub display_name: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Referee {
    pub referee_id: String,
    pub nba_api_ref_id: Option<i64>,
    pub display_name: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Game {
    pub game_id: String,
    pub league_id: i32,
    pub season_id: i32,
    pub season_type: String,
    pub game_date: String,
    pub home_team_id: i64,
    pub away_team_id: i64,
    pub home_points: Option<i32>,
    pub away_points: Option<i32>,
    pub attendance: Option<i32>,
    pub arena_id: Option<i32>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PlayerBoxscore {
    pub game_id: String,
    pub player_id: i64,
    pub team_id: i64,
    pub minutes: Option<f64>,
    pub points: Option<i32>,
    pub assists: Option<i32>,
    pub rebounds_total: Option<i32>,
    pub steals: Option<i32>,
    pub blocks: Option<i32>,
    pub fgm: Option<i32>,
    pub fga: Option<i32>,
    pub fg3m: Option<i32>,
    pub fg3a: Option<i32>,
    pub ftm: Option<i32>,
    pub fta: Option<i32>,
    pub pf: Option<i32>,
    pub turnovers: Option<i32>,
    pub plus_minus: Option<i32>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TeamBoxscore {
    pub game_id: String,
    pub team_id: i64,
    pub minutes: Option<f64>,
    pub points: Option<i32>,
    pub assists: Option<i32>,
    pub rebounds_total: Option<i32>,
    pub fgm: Option<i32>,
    pub fga: Option<i32>,
    pub fg3m: Option<i32>,
    pub fg3a: Option<i32>,
    pub ftm: Option<i32>,
    pub fta: Option<i32>,
    pub turnovers: Option<i32>,
    pub pf: Option<i32>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Draft {
    pub season_year: i32,
    pub pick_overall: i32,
    pub round_number: Option<i32>,
    pub pick_in_round: Option<i32>,
    pub team_abbrev: Option<String>,
    pub player_bref_id: Option<String>,
    pub player_name: Option<String>,
    pub college: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PlayerSeasonTotal {
    pub season_year: i32,
    pub player_bref_id: String,
    pub player_name: Option<String>,
    pub team_abbrev: String,
    pub games: i32,
    pub games_started: Option<i32>,
    pub minutes: Option<f64>,
    pub points: Option<i32>,
    pub assists: Option<i32>,
    pub rebounds_total: Option<i32>,
    pub steals: Option<i32>,
    pub blocks: Option<i32>,
    pub fgm: Option<i32>,
    pub fga: Option<i32>,
    pub fg3m: Option<i32>,
    pub fg3a: Option<i32>,
    pub ftm: Option<i32>,
    pub fta: Option<i32>,
    pub turnovers: Option<i32>,
    pub pf: Option<i32>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PlayerSeasonAdvanced {
    pub season_year: i32,
    pub player_bref_id: String,
    pub team_abbrev: String,
    pub minutes: Option<f64>,
    pub per: Option<f64>,
    pub ts_percent: Option<f64>,
    pub usg_percent: Option<f64>,
    pub ows: Option<f64>,
    pub dws: Option<f64>,
    pub ws: Option<f64>,
    pub bpm: Option<f64>,
    pub vorp: Option<f64>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TeamSeasonTotal {
    pub season_year: i32,
    pub team_abbrev: String,
    pub games: i32,
    pub minutes: Option<f64>,
    pub points: Option<i32>,
    pub assists: Option<i32>,
    pub rebounds_total: Option<i32>,
    pub fgm: Option<i32>,
    pub fga: Option<i32>,
    pub fg3m: Option<i32>,
    pub fg3a: Option<i32>,
    pub ftm: Option<i32>,
    pub fta: Option<i32>,
    pub turnovers: Option<i32>,
    pub pf: Option<i32>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TeamSeasonAdvanced {
    pub season_year: i32,
    pub team_abbrev: String,
    pub wins: i32,
    pub losses: i32,
    pub srs: Option<f64>,
    pub pace: Option<f64>,
    pub off_rtg: Option<f64>,
    pub def_rtg: Option<f64>,
    pub net_rtg: Option<f64>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Standing {
    pub season_year: i32,
    pub team_abbrev: String,
    pub wins: i32,
    pub losses: i32,
    pub playoffs: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Award {
    pub award_key: String,
    pub award_type: String,
    pub season_year: i32,
    pub player_bref_id: Option<String>,
    pub player_name: Option<String>,
    pub team_abbrev: Option<String>,
    pub rank: Option<i32>,
    pub points_won: Option<i32>,
    pub points_max: Option<i32>,
    pub share: Option<f64>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Transaction {
    pub transaction_id: String,
    pub season_year: i32,
    pub team_abbrev: Option<String>,
    pub player_bref_id: Option<String>,
    pub details: String,
}

File: convex/schema.ts
import { defineSchema, defineTable } from "convex/server";
import { v } from "convex/values";

export default defineSchema({
  leagues: defineTable({
    league_id: v.number(),
    league_code: v.string(),
    league_name: v.string(),
  }).index("by_league_id", ["league_id"]),

  seasons: defineTable({
    season_id: v.number(),
    league_id: v.number(),
    season_year: v.number(),
    start_date: v.optional(v.string()),
    end_date: v.optional(v.string()),
  })
    .index("by_season_id", ["season_id"])
    .index("by_league_year", ["league_id", "season_year"]),

  arenas: defineTable({
    arena_id: v.number(),
    arena_name: v.string(),
    city: v.optional(v.string()),
  }).index("by_arena_id", ["arena_id"]),

  teams: defineTable({
    team_id: v.number(), // BIGINT in SQL
    league_id: v.number(),
    franchise_code: v.optional(v.string()),
    nba_api_team_id: v.optional(v.number()),
  }).index("by_team_id", ["team_id"]),

  team_history: defineTable({
    team_history_id: v.number(),
    team_id: v.number(),
    effective_start: v.string(), // DATE
    effective_end: v.optional(v.string()), // DATE
    city: v.string(),
    nickname: v.string(),
    abbreviation: v.optional(v.string()),
    is_active: v.boolean(),
  })
    .index("by_team_history_id", ["team_history_id"])
    .index("by_team_id", ["team_id"])
    .index("by_abbreviation", ["abbreviation"])
    .index("by_active", ["is_active", "team_id"]),

  players: defineTable({
    player_id: v.number(), // BIGINT
    nba_api_person_id: v.optional(v.number()),
    first_name: v.optional(v.string()),
    last_name: v.optional(v.string()),
    display_name: v.optional(v.string()),
    birth_date: v.optional(v.string()),
    from_year: v.optional(v.number()),
    to_year: v.optional(v.number()),
  })
    .index("by_player_id", ["player_id"])
    .index("by_last_name", ["last_name", "first_name"]),

  coaches: defineTable({
    coach_id: v.string(),
    nba_api_coach_id: v.optional(v.number()),
    display_name: v.string(),
  }).index("by_coach_id", ["coach_id"]),

  referees: defineTable({
    referee_id: v.string(),
    nba_api_ref_id: v.optional(v.number()),
    display_name: v.string(),
  }).index("by_referee_id", ["referee_id"]),

  games: defineTable({
    game_id: v.string(),
    league_id: v.number(),
    season_id: v.number(),
    season_type: v.string(),
    game_date: v.string(), // DATE
    home_team_id: v.number(),
    away_team_id: v.number(),
    home_points: v.optional(v.number()),
    away_points: v.optional(v.number()),
    attendance: v.optional(v.number()),
    arena_id: v.optional(v.number()),
  })
    .index("by_game_id", ["game_id"])
    .index("by_season_id", ["season_id"])
    .index("by_home_team", ["home_team_id"])
    .index("by_away_team", ["away_team_id"]),

  player_boxscores: defineTable({
    game_id: v.string(),
    player_id: v.number(),
    team_id: v.number(),
    minutes: v.optional(v.number()), // DOUBLE
    points: v.optional(v.number()),
    assists: v.optional(v.number()),
    rebounds_total: v.optional(v.number()),
    steals: v.optional(v.number()),
    blocks: v.optional(v.number()),
    fgm: v.optional(v.number()),
    fga: v.optional(v.number()),
    fg3m: v.optional(v.number()),
    fg3a: v.optional(v.number()),
    ftm: v.optional(v.number()),
    fta: v.optional(v.number()),
    pf: v.optional(v.number()),
    turnovers: v.optional(v.number()),
    plus_minus: v.optional(v.number()),
  })
    .index("by_game_id", ["game_id"])
    .index("by_player_id", ["player_id"])
    .index("by_game_player", ["game_id", "player_id"]),

  team_boxscores: defineTable({
    game_id: v.string(),
    team_id: v.number(),
    minutes: v.optional(v.number()),
    points: v.optional(v.number()),
    assists: v.optional(v.number()),
    rebounds_total: v.optional(v.number()),
    fgm: v.optional(v.number()),
    fga: v.optional(v.number()),
    fg3m: v.optional(v.number()),
    fg3a: v.optional(v.number()),
    ftm: v.optional(v.number()),
    fta: v.optional(v.number()),
    turnovers: v.optional(v.number()),
    pf: v.optional(v.number()),
  })
    .index("by_game_id", ["game_id"])
    .index("by_team_id", ["team_id"])
    .index("by_game_team", ["game_id", "team_id"]),

  player_season_totals: defineTable({
    season_year: v.number(),
    player_bref_id: v.string(),
    player_name: v.optional(v.string()),
    team_abbrev: v.string(),
    games: v.number(),
    games_started: v.optional(v.number()),
    minutes: v.optional(v.number()),
    points: v.optional(v.number()),
    assists: v.optional(v.number()),
    rebounds_total: v.optional(v.number()),
    steals: v.optional(v.number()),
    blocks: v.optional(v.number()),
    fgm: v.optional(v.number()),
    fga: v.optional(v.number()),
    fg3m: v.optional(v.number()),
    fg3a: v.optional(v.number()),
    ftm: v.optional(v.number()),
    fta: v.optional(v.number()),
    turnovers: v.optional(v.number()),
    pf: v.optional(v.number()),
  })
    .index("by_player_season_team", ["player_bref_id", "season_year", "team_abbrev"])
    .index("by_player_season", ["player_bref_id", "season_year"])
    .index("by_season_year", ["season_year"])
    .index("by_season_team", ["season_year", "team_abbrev"]),

  player_season_advanced: defineTable({
    season_year: v.number(),
    player_bref_id: v.string(),
    team_abbrev: v.string(),
    minutes: v.optional(v.number()),
    per: v.optional(v.number()),
    ts_percent: v.optional(v.number()),
    usg_percent: v.optional(v.number()),
    ows: v.optional(v.number()),
    dws: v.optional(v.number()),
    ws: v.optional(v.number()),
    bpm: v.optional(v.number()),
    vorp: v.optional(v.number()),
  })
    .index("by_player_season_team", ["player_bref_id", "season_year", "team_abbrev"])
    .index("by_player_season", ["player_bref_id", "season_year"])
    .index("by_season_year", ["season_year"])
    .index("by_season_team", ["season_year", "team_abbrev"]),

  team_season_totals: defineTable({
    season_year: v.number(),
    team_abbrev: v.string(),
    games: v.number(),
    minutes: v.optional(v.number()),
    points: v.optional(v.number()),
    assists: v.optional(v.number()),
    rebounds_total: v.optional(v.number()),
    fgm: v.optional(v.number()),
    fga: v.optional(v.number()),
    fg3m: v.optional(v.number()),
    fg3a: v.optional(v.number()),
    ftm: v.optional(v.number()),
    fta: v.optional(v.number()),
    turnovers: v.optional(v.number()),
    pf: v.optional(v.number()),
  })
    .index("by_team_season", ["team_abbrev", "season_year"])
    .index("by_season_year", ["season_year"]),

  team_season_advanced: defineTable({
    season_year: v.number(),
    team_abbrev: v.string(),
    wins: v.number(),
    losses: v.number(),
    srs: v.optional(v.number()),
    pace: v.optional(v.number()),
    off_rtg: v.optional(v.number()),
    def_rtg: v.optional(v.number()),
    net_rtg: v.optional(v.number()),
  })
    .index("by_team_season", ["team_abbrev", "season_year"])
    .index("by_season_year", ["season_year"]),

  standings: defineTable({
    season_year: v.number(),
    team_abbrev: v.string(),
    wins: v.number(),
    losses: v.number(),
    playoffs: v.boolean(),
  })
    .index("by_team_season", ["team_abbrev", "season_year"])
    .index("by_season_year", ["season_year"]),

  drafts: defineTable({
    season_year: v.number(),
    pick_overall: v.number(),
    round_number: v.optional(v.number()),
    pick_in_round: v.optional(v.number()),
    team_abbrev: v.optional(v.string()),
    player_bref_id: v.optional(v.string()),
    player_name: v.optional(v.string()),
    college: v.optional(v.string()),
  })
    .index("by_season_pick", ["season_year", "pick_overall"])
    .index("by_team_abbrev", ["team_abbrev"])
    .index("by_player_bref_id", ["player_bref_id"]),

  awards: defineTable({
    award_key: v.string(),
    award_type: v.string(),
    season_year: v.number(),
    player_bref_id: v.optional(v.string()),
    player_name: v.optional(v.string()),
    team_abbrev: v.optional(v.string()),
    rank: v.optional(v.number()),
    points_won: v.optional(v.number()),
    points_max: v.optional(v.number()),
    share: v.optional(v.number()),
  })
    .index("by_award_key", ["award_key"])
    .index("by_award_season", ["award_type", "season_year"])
    .index("by_season_year", ["season_year"])
    .index("by_player_bref_id", ["player_bref_id"]),

  transactions: defineTable({
    transaction_id: v.string(),
    season_year: v.number(),
    team_abbrev: v.optional(v.string()),
    player_bref_id: v.optional(v.string()),
    details: v.string(),
  })
    .index("by_transaction_id", ["transaction_id"])
    .index("by_season_year", ["season_year"])
    .index("by_player_bref_id", ["player_bref_id"]),
});
